porch/func/internal/podevaluator.go

```diff
@@ -65,30 +74,109 @@ func NewPodEvaluator(namespace, wrapperServerImage string) (Evaluator, error) {
-	}, nil
+		gcScanInternal:     time.Duration(interval) * time.Second,
+		podTTL:             time.Duration(ttl) * time.Second,
+		imageEntrypoint:    map[string][]string{},
+		grpcConnectionPool: map[string]*grpc.ClientConn{},
+	}
+	go pe.garbageCollector()
+	return pe, nil
 }
 
-func (pe *podEvaluator) EvaluateFunction(ctx context.Context, req *pb.EvaluateFunctionRequest) (*pb.EvaluateFunctionResponse, error) {
-	ref, err := name.ParseReference(req.Image)
-	if err != nil {
-		return nil, err
+// TODO: We can use Watch + periodically reconciliation to manage the pods,
+// the pod evaluator will become a controller.
+func (pe *podEvaluator) garbageCollector() {
+	var err error
+	podList := &corev1.PodList{}
+	for {
+		time.Sleep(pe.gcScanInternal)
+		err = pe.kubeClient.List(context.Background(), podList, client.InNamespace(pe.namespace))
+		if err != nil {
+			klog.Warningf("unable to list pods in namespace %v: %w", pe.namespace, err)
+			continue
+		}
+		for _, pod := range podList.Items {
+			lastUse, found := pod.Annotations[lastUseTimeAnnotation]
+			// If a pod doesn't have a last-use annotation, we patch it.
+			if !found {
+				if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+					klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+				}
+			} else {
+				lu, err := strconv.ParseInt(lastUse, 10, 64)
+				if err != nil {
+					klog.Warningf("unable to convert the Unix time string to int64: %w", err)
+					// If the annotation is ill-formatted, we patch it with the current time and will try to GC it later.
+					if err = patchPodWithUnixTimeAnnotation(context.Background(), pe.kubeClient, &pod); err != nil {
+						klog.Warningf("can't patch pod %v/%v with the last-use annotation: %w", pod.Namespace, pod.Name, err)
+					}
+					continue
+				}
+				if time.Unix(lu, 0).Add(pe.podTTL).Before(time.Now()) {
+					// deleting the grpc connection from the pool
+					image := pod.Spec.Containers[0].Image
+					cc, found := pe.grpcConnectionPool[image]
+					if found {
+						klog.Infof("deleting grpc connection %v for %v", cc.Target(), image)
+						delete(pe.grpcConnectionPool, image)
+					}
+					// deleting the old pod
+					klog.Infof("deleting pod %v/%v", pod.Namespace, pod.Name)
+					err = pe.kubeClient.Delete(context.Background(), &pod)
+					if err != nil {
+						klog.Warningf("unable to delete pod %v/%v: %w", pod.Namespace, pod.Name, err)
+						continue
+					}
+				}
+			}
+		}
 	}
-	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
-	if err != nil {
-		return nil, err
+}
+
+func (pe *podEvaluator) retrieveOrCreatePod(ctx context.Context, image string) (client.ObjectKey, error) {
+	// Try to retrieve the pod first.
+	// Lookup the pod by label to see if there is a pod that can be reused.
+	// Looking it up locally may not work if there are more than one instance of the function runner,
```

Comment:
although we should probably check local cache first, and if not found, check for live pods (created by another runner), and if not found, create a new pod?